### DISCUSSION TITLE
How to implement multitenancy using virtual hosts in emqx 5.026.

### DISCUSSION THREAD

**Discussion by sc-sityad** [emqx/emqx#10937] [Q&A]
How can I isolate the whole client, its data, configuration etc from other clients.

**Comment by lafirest** [comment]
There already is a multitenancy version of EMQX called `Serverless`, you can have a try [here](https://www.emqx.com/en/cloud/serverless-mqtt).

**Comment by sc-sityad** [comment]
It is paid one right Can I impliment it in open source emqx 5.0.26?

**Reply by lafirest** [reply]
Yes, of course, but it is not easy to explain how to implement this.
In short words, you have to deep into the source code and refactor the topic, the router and other data structures to distinguish different tenancies.

**Comment by zmstone** [comment]
Hi @sc-sityad 

Our multi-tenancy solution is not opensource.
But I can tell you that it's not based on virtual hosts.

Here are some high-level summary how we did it.

* SNI is used as tenant ID
* Tenant ID is used as a dynamic mount-point for topic namespace isolation
* Authentication and Authorization are based on topic namespaces

less important aspects:
* Metrics are namespaced
* Independent limiters for different tenants
